src/lib/raft/raft.c

```diff
@@ -775,17 +859,18 @@ raft_sm_schedule_new_election_cb(struct ev_loop *loop, struct ev_timer *timer,
 	 */
 	assert(raft_is_fully_on_disk(raft));
 	raft_ev_timer_stop(loop, timer);
-	raft_sm_schedule_new_election(raft);
+	bool was_seen = raft_leader_unsee(raft);
+	if (!was_seen)
+		raft_sm_election_update(raft);
```

Comment:
I've been looking at this for at least an hour in total I think, still looks really nasty. That we call this update only because it is not always called inside of `unsee()`. Maybe inline `leader_unsee` here without the update and do the update in the end regardless of what was the bit? To emphasize that we are doing the update regardless of anything.

Reply:
Agree. It may be simplified quite a lot.
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index d0b26c3e3..9e724f334 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -858,9 +858,9 @@ raft_sm_election_update_cb(struct ev_loop *loop, struct ev_timer *timer,
         */
        assert(raft_is_fully_on_disk(raft));
        raft_ev_timer_stop(loop, timer);
-       bool was_seen = raft_leader_unsee(raft);
-       if (!was_seen)
-               raft_sm_election_update(raft);
+       bit_clear(&raft->leader_witness_map, raft->self);
+       raft_schedule_broadcast(raft);
+       raft_sm_election_update(raft);
 }
 
 static void
```